localstack/services/generic_proxy.py

```diff
@@ -592,136 +594,88 @@ def get_flask_ssl_context(cls, serial_number=None):
-        #  forwarding if not enabled
-        if self.base_url.netloc:
-            stripped_netloc, _, port = self.base_url.netloc.rpartition(":")
-            if host != self.base_url.netloc and (
-                host != stripped_netloc or port not in ["80", "443"]
-            ):
-                return False
-
-        # check path components
-        if self.base_url.path == "/":
-            if path.startswith("/"):
-                return True
-
-        path_parts = path.split("/")
-        base_path_parts = self.base_url.path.split("/")
-
-        if len(base_path_parts) > len(path_parts):
-            return False
-
-        for i, component in enumerate(base_path_parts):
-            if component != path_parts[i]:
-                return False
-
-        return True
-
-    def build_forward_url(self, host, path):
-        # build forward url
-        if self.forward_url.hostname:
-            forward_host = self.forward_url.scheme + "://" + self.forward_url.netloc
-        else:
-            forward_host = host
-        forward_path_root = self.forward_url.path
-        forward_path = path[len(self.base_url.path) :]  # strip base path
-
-        # avoid double slashes
-        if forward_path and not forward_path_root.endswith("/"):
-            if not forward_path.startswith("/"):
-                forward_path = "/" + forward_path
-
-        forward_url = forward_host + forward_path_root + forward_path
+    def register(self, forward_url):
+        """
+        Registers two routes at the edge router to implement either a path or a domain based routing.
+        We need two rules to match 1) the root path + no additional path 2) the root path + any other path.
+        This function is NOT called automatically, it is usually called in the leaf class' start up routine.
 
-        return urlparse(forward_url)
+        :param forward_url: The url which the proxy forwards to (from the initially passed base_url)
+        """
 
+        #   We MUST NOT create a catch all traffic rule
+        assert not (
+            (self._url.path == "" or self._url.path is None or self._url.path == "/")
+            and self._url.hostname == config.LOCALSTACK_HOSTNAME
+        )
+        from localstack.services.edge import ROUTER
 
-class EndpointProxy(ProxyListener):
-    def __init__(self, base_url: str, forward_url: str) -> None:
-        super().__init__()
-        self.forwarder = UrlMatchingForwarder(
-            base_url=base_url,
-            forward_url=forward_url,
+        self.rules.append(
+            ROUTER.add(
+                self._url.path or "/",
+                ProxyHandler(forward_url),
+                f"{self._url.hostname}<regex('(:.*)?'):port>",
+            )
+        )
+        self.rules.append(
+            ROUTER.add(
+                f"{self._url.path or ''}/<path:path>",
+                ProxyHandler(forward_url),
+                f"{self._url.hostname}<regex('(:.*)?'):port>",
```

Comment:
Unfortunately, you cannot always add a host rule in addition to a path rule.
If you add a host rule which checks for "localhost", the router will not match if the `Host` header contains anything else than localhost.